source/geometry_model/box.cc

```diff
@@ -260,6 +260,183 @@ namespace aspect
+                                          update_quadrature_points);
+
+
+        // Loop over all corners at the surface and save their X and Y positions.
+        // TODO: Update this to work in 3D. Spherical?
+        for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+          if (cell->is_locally_owned() && cell->at_boundary())
+            for (const unsigned int face_no : cell->face_indices())
+              if (cell->face(face_no)->at_boundary())
+                {
+                  if ( cell->face(face_no)->boundary_id() != relevant_boundary)
+                    continue;
+
+                  fe_face_values.reinit(cell, face_no);
+
+                  for (unsigned int corner = 0; corner < face_corners.size(); ++corner)
+                    {
+                      const Point<dim> vertex = fe_face_values.quadrature_point(corner);
+
+                      local_surface_height[0].push_back(vertex(0));   // X
+                      local_surface_height[1].push_back(vertex(dim-1));  // Y, now this only works for 2D
+
+                    }
+                }
+
+        // Combine all local_surfaces, combine and sort them, and broadcast back.
+        if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)
+        {
+          // Only push back them if the x value changes. This is done because
+          // there are two corner points for each position, and we want to remove duplicates.
+          // TODO: Is there an easier way to remove duplicates before getting here?
+          for (unsigned int i=0; i<local_surface_height[1].size(); i++)
+          {
+            if(i==0)
+            {
+              temp_surface_x.push_back(local_surface_height[0][i]);
+              temp_surface_y.push_back(local_surface_height[1][i]);
+            }
+            else
+              if(local_surface_height[0][i] != local_surface_height[0][i-1])
+              {
+                 temp_surface_x.push_back(local_surface_height[0][i]);
+                 temp_surface_y.push_back(local_surface_height[1][i]);
+              }
+          }
+
+          for (unsigned int p=1; p<Utilities::MPI::n_mpi_processes(this->get_mpi_communicator()); ++p)
+          {
+            // First, find out the size of the array a process wants to send.
+            MPI_Status status;
+            MPI_Probe(p, 42, this->get_mpi_communicator(), &status);
+            int incoming_size = 0;
+            MPI_Get_count(&status, MPI_DOUBLE, &incoming_size);
+
+            // Resize the array so it fits whatever the process sends.
+            std::vector<std::vector<double>> temporary_surface(2, std::vector<double>());
+
+            for (unsigned int i=0; i<temporary_surface.size(); ++i)
+                temporary_surface[i].resize(incoming_size);
+
+            for (unsigned int i=0; i<temporary_surface.size(); ++i)
+              MPI_Recv(&temporary_surface[i][0], incoming_size, MPI_DOUBLE, p, 42, this->get_mpi_communicator(), &status);
```

Comment:
There are easier ways to do this. Here, you want to use this function: https://dealii.org/developer/doxygen/deal.II/namespaceUtilities_1_1MPI.html#ad45956802191bd6765cdc5291a605acf

Reply:
The point being that deal.II's isend/irecv/broadcast functions can deal with arbitrary object types. You won't have to worry about it being a vector of doubles, for example. It can be more or less any data structure you choose, however complicated -- may be a `std::map`, for example.